libbpf-tools/memleak.c

```diff
@@ -0,0 +1,707 @@
+				stack = &stacks[i];
+				break;
+			}
+			if (stacks[i].stack_id == 0) {
+				stack = &stacks[i];
+				stack->stack_id = info.stack_id;
+				cnt++;
+				break;
+			}
+		}
+		if (stack == NULL) continue;
+		stack->size += info.size;
+		stack->nr++;
+		if (env.show_allocs) {
+			printf("\taddr = %lx size = %lld\n", addr, info.size);
+		}
+	}
+
+	quick_sort(stacks, 0, cnt);
+	if (cnt < max_prints) {
+		max_prints = cnt;
+	}
+
+	for (i=0; i < max_prints; i++) {
+		if (stacks[i].stack_id == 0) break;
+
+		if (bpf_map_lookup_elem(sfd, &stacks[i].stack_id, ip) != 0) {
+			fprintf(stderr, "    [Missed User Stack]\n");
+			continue;
+		}
+
+		if (env.kernel_trace) {
+			printf("\t%u bytes in %u allocations from stack\n", stacks[i].size, stacks[i].nr);
+			for (int j = 0; j < env.max_stack_depth && ip[j]; j++) {
+				ksym = ksyms__map_addr(ksyms, ip[j]);
+				if (ksym)
+					printf("\t\t%s (0x%lx)\n", ksym->name, ksym->addr);
+				else
+					printf("\t\t[unknown]\n");
+			}
+
+		} else {
+			syms = syms_cache__get_syms(syms_cache, env.pid);
+			if (!syms) {
+				fprintf(stderr, "failed to get syms\n");
+				continue;
+			}
+
+			printf("\t%u bytes in %u allocations from stack\n", stacks[i].size, stacks[i].nr);
+			for (int j = 0; j < env.max_stack_depth && ip[j]; j++) {
+				sym = syms__map_addr(syms, ip[j]);
+				if (sym)
+					printf("\t\t%s (0x%lx)\n", sym->name, sym->start);
+				else
+					printf("\t\t[unknown]\n");
+			}
+		}
+	}
+
+cleanup:
+	if (ip) free(ip);
+	if (stacks) free(stacks);
+}
+
+void sig_handler(int signo)
+{
+	if (env.child_pid == 0)
+		return;
+
+	if (signo == SIGINT || signo == SIGTERM) {
+		// kill all child processes forked by command
+		kill(0, SIGKILL);
+	}
+}
+
+int main(int argc, char **argv)
+{
+	char libc_path[PATH_MAX] = {};
+	struct memleak_bpf *skel;
+	struct syms_cache *syms_cache = NULL;
+	struct ksyms *ksyms = NULL;
+	int err, i;
+
+	/* Parse command line arguments */
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	if (env.command != NULL) {
+		const char *delim = " ";
+		char *cmd = strdup(env.command);
+		char **argv, *ptr, *filepath;
+		int j;
+		argv = malloc(sizeof(char *) * strlen(cmd));
+		memset(argv, 0, sizeof(char *) * strlen(cmd));
+		ptr = strtok(cmd, delim);
+		if (ptr != NULL) {
+			filepath = ptr;
+			ptr = strtok(NULL, delim);
+		} else {
+			fprintf(stderr, "Failed to exec %s\n", env.command);
+			exit(-1);
+		}
+
+		j = 0;
+		while (ptr != NULL)
+		{
+			argv[j++] = ptr;
+			ptr = strtok(NULL, delim);
+		}
+
+		env.pid = fork();
+		if (env.pid == 0) {
+			execve(filepath, argv, NULL);
+		} else if (env.pid > 0) {
+			// main process
+			env.child_pid = env.pid;
+			signal(SIGINT, sig_handler);
+		} else {
+			fprintf(stderr, "Failed to exec %s\n", env.command);
+			exit(-1);
+		}
+	}
+	env.kernel_trace = env.pid == 0;
+
+	/* Set up libbpf errors and debug info callback */
+	if (env.trace_all) {
+		libbpf_set_print(libbpf_print_fn);
+	}
```

Comment:
Please refer to other tools.

`env.trace_all` should be used in BPF program only, no ?

Reply:
Fix in commit [2beb289e](https://github.com/iovisor/bcc/pull/3828/commits/2beb289e98d651bddc2127299c5627b5d6149fce).